op-node/benchmarks/batchbuilding_test.go

```diff
@@ -0,0 +1,128 @@
+		EpochNum:     rollup.Epoch(1 + rng.Int63n(100_000_000)),
+		EpochHash:    testutils.RandomHash(rng),
+		Timestamp:    uint64(rng.Int63n(2_000_000_000)),
+		Transactions: txsEncoded,
+	}
+}
+
+type BatchingBenchmarkTC struct {
+	BatchType  uint
+	BatchCount int
+	txPerBatch int
+}
+
+func (t BatchingBenchmarkTC) String(cName string) string {
+	var btype string
+	if t.BatchType == derive.SingularBatchType {
+		btype = "Singular"
+	}
+	if t.BatchType == derive.SpanBatchType {
+		btype = "Span"
+	}
+	return fmt.Sprintf("BatchType=%s, txPerBatch=%d, BatchCount=%d, Compressor=%s", btype, t.txPerBatch, t.BatchCount, cName)
+}
+
+// BenchmarkChannelOut benchmarks the performance of adding singular batches to a channel out
+// this exercises the compression and batching logic, as well as any batch-building logic
+// Every Compressor in the compressor map is benchmarked for each test case
+func BenchmarkChannelOut(b *testing.B) {
+	rc, _ := compressor.NewRatioCompressor(compressor.Config{
+		TargetFrameSize:  100000,
+		TargetNumFrames:  1,
+		ApproxComprRatio: 0.4,
+	})
+	sc, _ := compressor.NewShadowCompressor(compressor.Config{
+		TargetFrameSize:  100000,
+		TargetNumFrames:  1,
+		ApproxComprRatio: 0.4,
+	})
+	nc, _ := compressor.NewNonCompressor(compressor.Config{
+		TargetFrameSize:  100000,
+		TargetNumFrames:  1,
+		ApproxComprRatio: 0.4,
+	})
+
+	compressors := map[string]derive.Compressor{
+		"RatioCompressor":  rc,
+		"ShadowCompressor": sc,
+		"NonCompressor":    nc,
+	}
+
+	tests := []BatchingBenchmarkTC{
+		// Singular Batch Tests
+		// low-throughput chains
+		{derive.SingularBatchType, 10, 1},
+		{derive.SingularBatchType, 50, 1},
+		{derive.SingularBatchType, 100, 1},
+		{derive.SingularBatchType, 200, 1},
+		{derive.SingularBatchType, 1000, 1},
+
+		// higher-throughput chains
+		{derive.SingularBatchType, 10, 10},
+		{derive.SingularBatchType, 100, 10},
+
+		// Span Batch Tests
+		// low-throughput chains
+		{derive.SpanBatchType, 10, 1},
+		{derive.SpanBatchType, 50, 1},
+		{derive.SpanBatchType, 100, 1},
+		{derive.SpanBatchType, 200, 1},
+		{derive.SpanBatchType, 1000, 1},
+
+		// higher-throughput chains
+		{derive.SpanBatchType, 10, 10},
+		{derive.SpanBatchType, 100, 10},
+	}
+
+	// for each compressor, run each the tests
+	for cName, c := range compressors {
+		for _, tc := range tests {
+			chainID := big.NewInt(333)
+			spanBatchBuilder := derive.NewSpanBatchBuilder(0, chainID)
+			rng := rand.New(rand.NewSource(0x543331))
+			c.Reset()
+			// pre-generate batches to keep the benchmark from including the random generation
+			batches := make([]*derive.SingularBatch, tc.BatchCount)
+			for i := 0; i < tc.BatchCount; i++ {
+				batches[i] = RandomSingularBatch(rng, tc.txPerBatch, chainID)
+			}
+			b.Run(tc.String(cName), func(b *testing.B) {
+				cout, _ := derive.NewChannelOut(tc.BatchType, c, spanBatchBuilder)
+				for i := 0; i < tc.BatchCount; i++ {
+					cout.AddSingularBatch(batches[i], 0)
```

Comment:
we should test here that it's not full, at least in a test run. otherwise it won't add anything after being full and the benchmark would be distorted.

Reply:
If it's full, we would stop that test-case early? If so, doesn't that distort the test in the same way that repeatedly hitting Full-Err would?

(I added this check anyway, as it makes sense to do so)

Reply:
it should just never be full. if we hit full, we need to increase the TOS

Reply:
FYI, this `derive` const needs to be raised in order to avoid hitting full channels in this test:
```
const MaxRLPBytesPerChannel = 10_000_000_000
```